src/Node.hs

```diff
@@ -89,27 +120,62 @@ startNode transport prng workers listeners = do
       nodeWorkers = tids
     }
   where
-    handlerIn :: NodeId -> ChannelIn m -> m ()
-    handlerIn peer chan = return ()
-    --TODO: fill in the dispatcher impl:
-    --      it needs to decode the MessageName
-    --      then lookup the listener(s)
-    --      then decode the body and fork the listener with the decoded message
+    -- Index the listeners by message name, for faster lookup.
+    listenerIndex :: ListenerIndex m
+    (listenerIndex, conflictingNames) = makeListenerIndex listeners
+
+    handlerIn :: LL.Node m -> SendActions m -> NodeId -> ChannelIn m -> m ()
+    handlerIn node sendActions peerId (ChannelIn chan) = do
+        (msgName :: MessageName, rest) <- recvPart chan (fromString "")
+        let listener = M.lookup msgName listenerIndex
+        case listener of
+            Just (ListenerActionOneMsg action) -> do
+                (msgBody, rest') <- recvPart chan rest
+                tid <- fork (action peerId sendActions msgBody)
+                -- TODO remember the thread id? Inform dispatcher when it's
+                -- finished?
+                pure ()
+            -- If it's a conversation listener, then that's an error, no?
+            Just (ListenerActionConversation _) -> error ("Wrong listener type! " ++ show msgName)
```

Comment:
Seems to me like it's better to use `throw` here since we have `Mockable Throw m` constraint.
But yes, I agree, that here should be an error if we encounter conversation listener according to my understanding of model.